packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -28,64 +26,84 @@ export class EntityHierarchyCacher {
-    // on startup, or when two projects share a hierarchy (at time of writing none do, but db schema
-    // makes it possible)
-    const alreadyCached = !!(await this.models.ancestorDescendantRelation.findOne({
-      entity_hierarchy_id: hierarchyId,
-      ancestor_id: parentId,
-    }));
-    if (alreadyCached) {
+  async fetchAndCacheDescendants(hierarchyId, parentIdsToAncestorIds) {
+    const parentIds = Object.keys(parentIdsToAncestorIds);
+
+    // check if we've reached the leaf nodes or this is already cached
+    const shouldFetchAndCache = await this.checkShouldFetchAndCache(hierarchyId, parentIds);
+    if (!shouldFetchAndCache) {
       return;
     }
 
-    // get the next generation of entities
-    const childIds = await this.getNextGeneration(hierarchyId, parentId);
+    // check whether next generation uses entity relation links, or should fall back to parent_id
+    const nextGenerationIsCanonical = await this.checkIfNextGenerationIsCanonical(
+      hierarchyId,
+      parentIds,
+    );
 
-    // if childIds is empty, we've made it to the leaf nodes
-    if (childIds.length === 0) {
-      return;
-    }
+    const childIdsToAncestorIds = {};
+    for (const parentId of parentIds) {
+      const ancestorIds = parentIdsToAncestorIds[parentId];
+      const childIds = nextGenerationIsCanonical
+        ? await this.getNextGenerationCanonically(parentId)
+        : await this.getNextGenerationViaEntityRelation(hierarchyId, parentId);
 
-    // cache this generation
-    const parentAndAncestorIds = [parentId, ...ancestorIds];
-    await this.cacheGeneration(hierarchyId, childIds, parentAndAncestorIds);
+      // if childIds is empty, we've made it to the leaf nodes
+      if (childIds.length === 0) {
+        continue;
+      }
+
+      // cache this generation
+      const parentAndAncestorIds = [parentId, ...ancestorIds];
+      await this.cacheGeneration(hierarchyId, childIds, parentAndAncestorIds);
+
+      // add this generation to the object for caching in the next recursion
+      childIds.forEach(childId => {
+        childIdsToAncestorIds[childId] = parentAndAncestorIds;
+      });
+    }
 
     // keep recursing through the hierarchy
-    const taskQueue = new AsyncTaskQueue(BATCH_SIZE);
-    const childTasks = childIds.map(childId =>
-      taskQueue.add(() =>
-        this.recursivelyFetchAndCacheDescendants(hierarchyId, childId, parentAndAncestorIds),
-      ),
-    );
-    await Promise.all(childTasks);
+    await this.fetchAndCacheDescendants(hierarchyId, childIdsToAncestorIds);
   }
 
-  async getNextGeneration(hierarchyId, entityId) {
-    // get any matching alternative hierarchy relationships leading out of these parents
-    const hierarchyLinks = await this.models.entityRelation.find({
-      parent_id: entityId,
+  async checkShouldFetchAndCache(hierarchyId, parentIds) {
+    if (parentIds.length === 0) {
+      return false; // base case of recursion, we must have reached the leaf nodes
```

Comment:
Perhaps the reference to recursion doesn't make sense now, but I'll leave this decision to you!

Reply:
I still find it useful :-)